Autoloader: Update version selection

kbrown9 opened the pull request:
#### Changes proposed in this Pull Request:
* Add a new class, `Version_Selector`, which is used by the `Autoloader_Handler`, `Classes_Handler`, and `Files_Handler` classes to select package versions.

#### Is this a new feature or does it add/remove features to an existing part of Jetpack?
* This changes the existing Autoloader package.

#### Testing instructions:
This PR updates the selection of package versions. Testing should cover the following scenarios:
 * Two or more plugins use different stable versions of a package. The latest version should be selected.
 * One plugin uses a stable version of a package, and another uses a development version of the package. The `JETPACK_AUTOLOAD_DEV` constant is not defined. The stable version should be used.
 * One plugin uses a stable version of a package, and another uses a development version of the package. The `JETPACK_AUTOLOAD_DEV` is defined as `true`. The development version should be used.

#### Proposed changelog entry for your changes:
* n/a

zinigor commented on packages/autoloader/src/class-version-selector.php:
```diff
@@ -0,0 +1,92 @@
+	 * @return Boolean Returns true if the selected package version should be updated,
+	 *                 else false.
+	 */
+	public function update_package_version( $class, $selected_version, $compare_version, $plugin ) {
```
What do you say if we rename this method to something that indicates that it returns a boolean, but doesn't actually update the version. Like `is_version_update_required` , or something of the sort. With the current name I was expecting this method to actually do something instead.

kbrown9 replied:
I agree, and I'll change the method name.

kbrown9 replied:
Done in commit cd9bf72

zinigor commented on packages/autoloader/src/class-version-selector.php:
```diff
@@ -0,0 +1,92 @@
+			return true;
+		}
+
+		if ( version_compare( $selected_version, $compare_version, '<' ) ) {
```
Judging by the logic, here you're expecting the `$compare_version` to never be a development version, right? If so, we can get a false positive if `99999-dev` gets through the previous conditions.

kbrown9 replied:
Sorry - I didn't realize you were already looking at this PR, and I force pushed a few changes, including one that fixes this logic error. [This change](https://github.com/Automattic/jetpack/pull/15135/files#diff-ef9d1216cc594d83f995ac323d19b7b0R42) should properly handle situations where the package is a development version but the plugin is not.

zinigor replied:
Hahah, OK, thanks! I couldn't help myself, this is very interesting stuff!

zinigor commented on packages/autoloader/src/class-version-selector.php:
```diff
@@ -0,0 +1,92 @@
+	 */
+	private function is_plugin_version_dev( $plugin ) {
+		$dir         = explode( '/', $plugin )[0];
+		$plugin_data = get_plugins( "/$dir" );
```
What do you say we solely rely here on `WP_DEBUG` and/or `JETPACK__DEV_DEBUG` constants? I think that would make everyone's life a lot easier. Want dev autoloader versions - turn on `WP_DEBUG` :)

kbrown9 replied:
This is great idea, and I think that would definitely make everyone's life easier! 

I think that I'd prefer to use `JETPACK__DEV_DEBUG` alone, since production sites occasionally have `WP_DEBUG` set to true, and it seems less likely that they would have `JETPACK__DEV_DEBUG` set to true.

zinigor replied:
One thing that we need to keep in mind here is that `JETPACK__DEV_DEBUG` also turns off a lot of stuff, I didn't think about that right away. Using that constant will mean that we won't be able to test some things.

kbrown9 replied:
Good point; I didn't know about that.

Do you agree that we shouldn't use `WP_DEBUG` because some production sites set it to true? What do you think of using a new constant, something like `JETPACK_AUTOLOADER_DEV_DEBUG`?

zinigor replied:
Yes, I agree about `WP_DEBUG` and yes, I wanted to suggest the same thing - let's introduce a new constant, but let's make it a bit easier on the eyes, something like `JETPACK_AUTOLOAD_DEV`? Or maybe you can come up with a better/shorter one? :)

kbrown9 replied:
I couldn't think of a short but still descriptive name, so I just went with `JETPACK_AUTOLOAD_DEV` in commit ea25283

zinigor reviewed (commented):
I like where this is going, thank you! I left some comments for discussion.

kbrown9 committed:
Autoloader: add Version_Selector class

Add a new class, Version_Selector, which is used by the Autoloader_Handler,
Classes_Handler, and Files_Handler classes to select package versions.

Remove Plugins_Handler::create_map_path_array() and Plugins_Handler::get_active_plugins_paths()
because they're no longer used.
